pkg/dns/aws/dns.go

```diff
@@ -0,0 +1,248 @@
+	if len(domain) == 0 {
+		return fmt.Errorf("domain is required")
+	}
+	if len(target) == 0 {
+		return fmt.Errorf("target is required")
+	}
+
+	// Ensure the zone cache is up to date.
+	err := m.updateZoneCache()
+	if err != nil {
+		return fmt.Errorf("failed to update hosted zone cache: %v", err)
+	}
+
+	publicZone := m.findClusterPublicZone()
+	privateZone := m.findClusterPrivateZone()
+	if publicZone == nil {
+		return fmt.Errorf("failed to find public zone for cluster")
+	}
+	if privateZone == nil {
+		return fmt.Errorf("failed to find private zone for cluster")
+	}
+
+	// Find the target hosted zone of the load balancer attached to the service.
+	// TODO: cache it?
+	var targetHostedZoneID string
+	loadBalancers, err := m.elb.DescribeLoadBalancers(&elb.DescribeLoadBalancersInput{})
+	if err != nil {
+		return fmt.Errorf("failed to describe load balancers: %v", err)
+	}
+	for _, lb := range loadBalancers.LoadBalancerDescriptions {
+		logrus.Debugf("found load balancer: %s (DNS name: %s, zone: %s)", aws.StringValue(lb.LoadBalancerName), aws.StringValue(lb.DNSName), aws.StringValue(lb.CanonicalHostedZoneNameID))
+		if aws.StringValue(lb.CanonicalHostedZoneName) == target {
+			targetHostedZoneID = aws.StringValue(lb.CanonicalHostedZoneNameID)
+			break
+		}
+	}
+	if len(targetHostedZoneID) == 0 {
+		return fmt.Errorf("couldn't find hosted zone ID of target ELB with domain name %s", target)
+	}
+
+	// Configure alias records for both the public and private zones. Cache
+	// successful updates.
+	// TODO: handle the caching/diff detection in a better way.
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	for _, zoneID := range []string{aws.StringValue(publicZone.Id), aws.StringValue(privateZone.Id)} {
+		key := zoneID + domain + target
+		// Only process these once, for now.
+		if _, exists := m.updatedRecords[key]; exists {
+			continue
+		}
+		err := m.updateAlias(domain, zoneID, target, targetHostedZoneID)
+		if err != nil {
+			return fmt.Errorf("failed to update alias in zone %s: %v", zoneID, err)
+		}
+		m.updatedRecords[key] = struct{}{}
+	}
+	return nil
+}
+
+// updateAlias creates or updates an alias for domain in zoneID pointed at
+// target in targetHostedZoneID.
+func (m *Manager) updateAlias(domain, zoneID, target, targetHostedZoneID string) error {
+	resp, err := m.route53.ChangeResourceRecordSets(&route53.ChangeResourceRecordSetsInput{
+		HostedZoneId: aws.String(zoneID),
+		ChangeBatch: &route53.ChangeBatch{
+			Changes: []*route53.Change{
+				{
+					Action: aws.String("UPSERT"),
+					ResourceRecordSet: &route53.ResourceRecordSet{
+						Name: aws.String(domain),
+						Type: aws.String("A"),
+						AliasTarget: &route53.AliasTarget{
+							HostedZoneId:         aws.String(targetHostedZoneID),
+							DNSName:              aws.String(target),
+							EvaluateTargetHealth: aws.Bool(false),
+						},
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("couldn't update DNS record for in zone %s: %v", zoneID, err)
```

Comment:
You have two prepositions and only one prepositional object: "for in zone %s".

Reply:
Fixed